商品購入機能

shokoron opened the pull request:
#What
商品購入機能を作成

#Why
商品購入機能を実装するため



 ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://i.gyazo.com/6e5ec1b983d1923182dab2cf675daaab.gif

 ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://i.gyazo.com/42314db837d32ec788e5ed9b958d9018.gif

 ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://i.gyazo.com/13206dd900e066e85a4f104407ca98fc.gif

売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
https://i.gyazo.com/e8ee7e36976c84d96548f92774d54635.gif
 
売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
https://i.gyazo.com/852d1b91cdb9172f7c78e2108a1da130.gif

ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://i.gyazo.com/9fdbe1cc20e5b0110dfe9f06f002fcfc.gif

 ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://i.gyazo.com/4d06377b26ecb8418bc78f4f0f4a39f3.gif

配送先（トークン有り）のテストコード画像
https://i.gyazo.com/5df616f9a31acbcb5a16c8e4e9dcb5fd.png

shokoron committed:
Formオブジェクト

shokoron committed:
配送先情報のテスト

shokoron committed:
商品購入機能(クレジットカード情報)

koyamatakumi commented on app/models/order_address.rb:
```diff
@@ -0,0 +1,28 @@
+    validates :phone_number, format: { with: /\A\d{10,11}\z/, message: "is invalid. Input only number without hyphen and must be 10 or 11 digits" }
+    validates :postal_code, presence: true, format: { with: /\A[0-9]{3}-[0-9]{4}\z/, message: "is invalid. Include hyphen(-)" }
+    validates :token, presence: true
+  end
```
```ruby
    validates :postal_code, presence: true, format: { with: /\A[0-9]{3}-[0-9]{4}\z/, message: "is invalid. Include hyphen(-)" }
    validates :token, presence: true
```
上記のpresence: trueは消しましょう。
■理由
 with_options presence: true do ~ end内に記載されているため、２重で制限をかけていることになっています。

koyamatakumi replied:
修正いただきありがとうございます。
tokenは消していただけましたが、以下の記述が修正されておりません。「presence: true」を消しましょう。
```ruby
validates :postal_code, presence: true, format: { with: /\A[0-9]{3}-[0-9]{4}\z/, message: "is invalid. Include hyphen(-)" }
```

koyamatakumi commented on app/views/items/index.html.erb:
```diff
@@ -135,11 +135,11 @@
+               <span>Sold Out!!</span>
+            </div>
+            <% end %>
           <%# //商品が売れていればsold outを表示しましょう %>
```
**実装案内コメント（〜しましょう）が残っています**
「〜しましょう」で終わる実装案内コメントは実装完了後に削除しましょう。
企業提出時に実装案内コメントが残っていないようにするためです。

koyamatakumi commented on app/views/items/show.html.erb:
```diff
@@ -10,11 +10,11 @@
+         <span>Sold Out!!</span>
+         </div>
+     <% end %>
       <%# //商品が売れている場合は、sold outを表示しましょう %>
```
**実装案内コメント（〜しましょう）が残っています**
「〜しましょう」で終わる実装案内コメントは実装完了後に削除しましょう。
企業提出時に実装案内コメントが残っていないようにするためです。

koyamatakumi commented on app/views/items/show.html.erb:
```diff
@@ -26,13 +26,13 @@
        <%= link_to "削除", item_path(@item), data: {turbo_method: :delete}, class:"item-destroy" %>
-      <% elsif "商品が売れていない条件" %>
-        <%= link_to "購入画面に進む", "#" ,class:"item-red-btn"%>
+      <% else "商品が売れていない条件" %>
```
```ruby
<% else "商品が売れていない条件" %>
```
「 "商品が売れていない条件" 」は不要です。消しておきましょう。

koyamatakumi commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,84 @@
+  before do
+    user = FactoryBot.create(:user)
+    item = FactoryBot.create(:item)
+    @order_address = FactoryBot.build(:order_address, user_id: user.id, item_id: item.id, token: "tok_abcdefghijk00000000000000000")
```
```ruby
    @order_address = FactoryBot.build(:order_address, user_id: user.id, item_id: item.id, token: "tok_abcdefghijk00000000000000000")
```
spec/factories/order_addresses.rbでtoken情報を用意しているため、「token: "tok_abcdefghijk00000000000000000"」は消しておきましょう。

koyamatakumi commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,84 @@
+        @order_address.token = nil
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include("Token can't be blank")
+      end
```
以下のテストを追加しましょう。
```
・userが紐づいていない場合登録できない。
・itemが紐づいていない場合登録できない。
・
```

koyamatakumi reviewed (commented):
shokoron様、レビュー依頼フォームの送信ありがとうございます。 
コードを確認させていただき、コメントいたしましたのでご確認をお願いいたします。
修正後はコミット、プッシュを忘れずに行い、改めてレビュー依頼フォームの送信をお願い致します！

shokoron committed:
修正しました

テスト修正版(userが紐づいていない場合登録できない・itemが紐づいていない場合登録できない)
https://i.gyazo.com/d808c3badfc430a80b8d88c1e41fcfb1.png

koyamatakumi reviewed (commented):
shokoron様 
レビュー依頼フォームの送信ありがとうございます。 修正いただいた内容を確認させていただき、追加でコメントいたしましたのでご確認お願いいたします。 

修正後はコミット、プッシュを忘れずに行い、改めてレビュー依頼フォームの送信をお願い致します。着実にゴールに近づいています。ラストスパート頑張っていきましょう！

shokoron committed:
修正しました

テスト修正版
https://i.gyazo.com/966e277afe120c59c2a549eba06e88e1.png

koyamatakumi commented:
shokoron様様 レビュー依頼フォームの送信ありがとうございます！ 
**LGTM**とさせていただきます。ここまでの実装本当にお疲れさまでした！ ！
マージ後、ローカルのmain(master)ブランチに忘れずにプルしましょう。
### ■ マージ後の手順 
```ruby
 ローカルのCurrentBrunchをmain(master)に変更　→　FetchOrigin　→　PullOrigin 
``` 
その後、次の作業に取り組んでいただけますと幸いです。休憩もはさみつつ引き続き頑張っていきましょう！
